util/buildRelease/gen_release

```diff
@@ -178,122 +102,259 @@ if (exists($ENV{"CHPL_GEN_RELEASE_NO_CLONE"})) {
+        git_url = os.getenv(
+            "CHPL_HOME_REPOSITORY", "https://github.com/chapel-lang/chapel"
+        )
+        git_branch = os.getenv("CHPL_GEN_RELEASE_BRANCH", "main")
+        repo_cache_path = os.getenv("REPO_CACHE_PATH", "/missing")
+        git_commit = os.getenv("CHPL_GEN_RELEASE_COMMIT", None)
+
+        print(
+            f"[gen_release] Cloning the sources (repo: {git_url} branch: {git_branch})..."
+        )
+        cmd = (
+            [
+                "git",
+                "clone",
+                "--reference-if-able",
+                f"{repo_cache_path}/chapel.git",
+                "--branch",
+                git_branch,
+            ]
+            + (["--depth", "1"] if git_commit is None else [])
+            + [
+                git_url,
+                rootdir,
+            ]
+        )
+        run(*cmd)
+
+        if git_commit is not None:
+            print(f"[gen_release] Checking out revision {git_commit}...")
+            with cd(rootdir):
+                run("git", "reset", "--hard", git_commit)
+
+        with cd(rootdir):
+            print("[gen_release] Confirm final Git source version used.")
+            run("git", "--no-pager", "status")
+            run("git", "--no-pager", "log", "-1")
+
+            # write the git-sha while we have the .git directory
+            run(
+                "util/config/write-git-sha",
+                "frontend/lib/util",
+                "--build-version",
+                f"--chpl-home={rootdir}",
+            )
+            # use git archive to copy all the files under source control into the archive directory.
+            print("[gen_release] Creating build workspace with git archive...")
+
+            p = run(
+                "git", "archive", "--format=tar", "HEAD", stdout=subprocess.PIPE
+            )
+            with cd(archive_dir):
+                run(tar_executable, "-xf", "-", input=p.stdout)
+
+            # explicitly copy the BUILD_VERSION file and git-version.cpp files
+            print("[gen_release] Copying BUILD_VERSION file.")
+            shutil.copyfile(
+                "compiler/main/BUILD_VERSION",
+                pjoin(archive_dir, "compiler/main/BUILD_VERSION"),
+            )
+            print("[gen_release] Copying git-version.cpp")
+            shutil.copyfile(
+                "frontend/lib/util/git-version.cpp",
+                pjoin(archive_dir, "frontend/lib/util/git-version.cpp"),
+            )
+
+        if "CHPL_HOME" in os.environ:
+            resultdir = pjoin(os.getenv("CHPL_HOME"), "tar")
+        else:
+            resultdir = basetmpdir
+
+    with cd(archive_dir):
+
+        print("[gen_release] Creating the spec tests...")
+        make("spectests")
+
+        # Docs/man page must be built first so we can get rid of any extra files
+        # (chpldoc) with a clobber
+        print("[gen_release] Building the docs...")
```

Comment:
You might want to define your own custom `log` function so as not to have to write `[gen_release]` each time